First to Third Deliverable

EmmDguez opened the pull request.

alvarocueva committed:
Add README file

EmmDguez committed:
ft: first deliverable.

EmmDguez commented:
Urls:
http://localhost:3000/pokemon
http://localhost:3000/pokemon/59
valid ids:
10,12,59,100

FerVT commented on interface/controller/poke_controller.go:
```diff
@@ -0,0 +1,39 @@
+
+func (pI *pokeController) GetPokemon(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id, _ := strconv.Atoi(vars["id"])
```
Make sure to handle the error here

EmmDguez replied:
Added code to handle it.

FerVT commented on interface/presenter/poke_presenter.go:
```diff
@@ -0,0 +1,23 @@
+}
+
+func (pP *pokePresenter) ResponsePokemon(pokething *model.Pokemon) *model.Pokemon {
+	return pokething
```
Please choose a more representative variable name, and use camelCase for it

FerVT commented on interface/repository/poke_repository.go:
```diff
@@ -0,0 +1,34 @@
+	"errors"
+
+	"github.com/ToteEmmanuel/academy-go-q12021/domain/model"
+	csv_file_reader "github.com/ToteEmmanuel/academy-go-q12021/tools/reader"
```
Avoid snake_case when importing packages, import aliases should be all lowercase and words together, in this example it should be like : `csvfilereader`

EmmDguez replied:
Removed the snake_case.

FerVT commented on registry/poke_registry.go:
```diff
@@ -0,0 +1,26 @@
+package registry
```
This package seems a bit of an overkill. It's ok to create all of the required services and dependencies on the main function instead

EmmDguez replied:
I removed the overkill package.

FerVT commented on usecase/interactor/poke_interactor.go:
```diff
@@ -0,0 +1,37 @@
+)
+
+type pokeInteractor struct {
+	PokeRepo      repository.PokeRepository
```
It's ok to define these interfaces here on the same package you're planning to use them

EmmDguez replied:
Did some refactoring on the interfaces.

FerVT commented on interface/presenter/poke_presenter.go:
```diff
@@ -0,0 +1,23 @@
+	return pokething
+}
+
+func (pP *pokePresenter) ResponsePokemons(pokethings []*model.Pokemon) []*model.Pokemon {
```
Also, is this package necessary if you're not doing anything with data anyway in your functions??

EmmDguez replied:
Removed it as it really added nothing to logic.

FerVT commented:
Also, avoid naming your files with snake_case, Go usually names the files all lowercase and words together, for example `pokepresenter.go`, even if it looks a bit ugly.

The only exception to this are test files

EmmDguez committed:
FT: second deliverable.
Create a client to consume an external API.
Add An endpoint to consume the external API.
Update the endpoint made in the first deliverable to display the result as a JSON.
Refactor as per comments in first deliverable.

Behind:
CSV file, unit testing.

EmmDguez commented:
Sorry got some issues with my time for this one I am a little behind in the schedule.

FT: second deliverable.
Create a client to consume an external API.
Add An endpoint to consume the external API.
Update the endpoint made in the first deliverable to display the result as a JSON.
Refactor as per comments in first deliverable.

Behind:
CSV file, unit testing.

The pokemon/###/catch endpoint will go to the pokemon api and retrieve the fields I am managing in my csv from the api.

FerVT commented on controller/pokecontroller.go:
```diff
@@ -30,10 +31,28 @@ func (pI *pokeController) GetPokemon(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.Header().Add("Content-Type", "Applicaiton/Json")
```
Found a typo! `Application/Json`. You would like to check your handlers to see if you missed some more

EmmDguez replied:
Yeah I did!, fixed them on the third deliverable.

FerVT commented on usecase/interactor/pokeinteractor.go:
```diff
@@ -0,0 +1,149 @@
+	var err error
+	err = fetchBase(catchedPokemon, pI.RestClient)
+	if err != nil {
+		fmt.Println(err)
```
Is it ok if you just log the error and not return here?

EmmDguez replied:
Changed the way the error is handled with a channel and an error response.

FerVT commented on usecase/interactor/pokeinteractor.go:
```diff
@@ -0,0 +1,149 @@
+		waitgroup.Done()
+		fmt.Println("Fetched Base...")
+	}()
+	waitgroup.Wait()
```
You would probably like to think of a way of handling both errors you're receiving from your go routines, if both of your APIs fail, then you will only know about one failure instead of both. There's probably a way of doing this with channels

EmmDguez replied:
followed the advice and used a channel to handle the errors.

FerVT reviewed (commented):
So far, it looks good.
However, please add unit testing whenever you can. Remember that it is a requirement for the last deliverable(and this one) to add testing for it to be considered complete

EmmDguez committed:
Missed date.

EmmDguez committed:
ft: stable functionality, missing unit testing.

EmmDguez committed:
ft: third deliverable refactored and with testing.

EmmDguez commented:
@FerVT I think I've addressed all the comments you previously made to me and have completed the points in the readme.
Thanks again!
"/pokemon"
"/pokemon/{id:[0-9]+}"
"/pokemon/{id:[0-9]+}/catch"
"/pokemon/workers" needs queryParams :"{type:[a-z]+}" "{items:[0-9]+}" "{items_per_workers:[0-9]+}"

it's currently running on http://localhost:4001/pokemon/100 but it can be changed before running it on config.yml.

FerVT commented on infrastructure/datastore/cvsfilereader.go:
```diff
@@ -0,0 +1,152 @@
+	for key := range c.pokeMap {
+		keys = append(keys, int(key))
+	}
+	odd := false
```
you could simplify this assignment like `odd := typeStr == "odd"`

EmmDguez replied:
followed the advice and used this form.

FerVT commented on infrastructure/datastore/cvsfilereader.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+func (c *CsvPokeStorage) Save(pokemon *model.Pokemon) (*model.Pokemon, error) {
+	if c.pokeMap[pokemon.Id] != nil {
```
A more reliable approach to this would be `if value, ok := c.pokeMap[pokemon.Id]; ok { ... }`

EmmDguez committed:
fix: simplify boolean "odd" assignment.

EmmDguez committed:
fix: use "ok" response instead of "!= nil" in Save.

EmmDguez commented:
Merged changes to address both comments into main.
